core/src/replay_stage.rs

```diff
@@ -462,6 +473,57 @@ impl ReplayStage {
         );
     }
 
+    fn reset_duplicate_slots(
+        duplicate_slots_reset_receiver: &DuplicateSlotsResetReceiver,
+        descendants: &HashMap<Slot, HashSet<Slot>>,
+        progress: &mut ProgressMap,
+        bank_forks: &RwLock<BankForks>,
+    ) {
+        for duplicate_slot in duplicate_slots_reset_receiver.try_iter() {
+            Self::purge_unconfirmed_duplicate_slot(
+                duplicate_slot,
+                descendants,
+                progress,
+                bank_forks,
+            );
+        }
+    }
+
+    fn purge_unconfirmed_duplicate_slot(
+        duplicate_slot: Slot,
+        descendants: &HashMap<Slot, HashSet<Slot>>,
+        progress: &mut ProgressMap,
+        bank_forks: &RwLock<BankForks>,
+    ) {
+        error!("purging slot {}", duplicate_slot);
+        let empty = HashSet::new();
+        let slot_descendants = descendants.get(&duplicate_slot).unwrap_or(&empty);
```

Comment:
nits: can use `unwrap_or_default()`?

Reply:
unfortunately this is of type `&std::collections::HashSet<u64>`, which doesn't implement default :'(

Reply:
oh i see!